### ISSUE TITLE
await quiescence of `!Send` types

### ISSUE THREAD

**Issue by droundy** [tokio-rs/tokio#5407]
**Is your feature request related to a problem? Please describe.**

I would like to implement (or have) a RCU pointer type with zero overhead for reading. This could be achieved by having reads occur through a non-Send guard (which need not implement Drop!).  Since the guard can't live through an await point, if we know that all worker threads have reached await points, we will know that all non-Send objects have been freed, and thus any copies may be freed.

I think I could achieve this using the various monotonically increasing counters per worker, but that seems awkward at best.  It would be nice to have a function like

```
/// Returns when all worker threads have dropped `! Send ` values
async fn await_quiescence() {...}
```

**Describe alternatives you've considered**

The other option is to track the `worker_poll_count()`? of every worker and return when they have all increased.  I'm not sure how to wait on this without polling, and not quite sure which counts I should wait on.

**Comment by droundy** [comment]
Possibly `wait_for_all_workers_to_be_polled` would be a better name?

My thinking in terms of this being part of tokio is that there might be a general idiom of creating `! Send` types and using this approach to determine if there all been freed.

**Comment by droundy** [comment]
I'm realizing that this idea relies on being able to ensure that my type is only used within a single tokio runtime. Is there a way to do this?

**Comment by Darksonn** [comment]
You cannot assume that non-Send values are dropped when you reach an await. They could have been stored in a thread-local, or the task might check the task id to figure out whether you're still on the same thread or not.

**Comment by droundy** [comment]
Good points.
